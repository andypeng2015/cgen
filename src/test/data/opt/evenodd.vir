module evenodd

function w $abs(w %n) {
@start:
  %c = slt.w %n, 0_w
  br %c, @neg, @pos
@neg:
  %n = neg.w %n
  ret %n
@pos:
  ret %n
}

export function b $even(w %n) {
@start:
  %n = call.w $abs(%n)
  %c = eq.w %n, 0_w
  br %c, @zero, @nonzero
@zero:
  ret 1_b
@nonzero:
  %n = sub.w %n, 1_w
  %b = call.b $odd(%n)
  ret %b
}

export function b $odd(w %n) {
@start:
  %n = call.w $abs(%n)
  %c = eq.w %n, 0_w
  br %c, @zero, @nonzero
@zero:
  ret 0_b
@nonzero:
  %n = sub.w %n, 1_w
  %b = call.b $even(%n)
  ret %b
}
