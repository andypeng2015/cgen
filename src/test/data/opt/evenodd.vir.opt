module evenodd

function w $abs(w %n) {
@2:
  %9 = slt.w %n, 0x0_w ; @30
  br %9, @3, @21(%n)
@3:
  %10 = neg.w %n ; @31
  jmp @21(%10)
@21(%0):
  ret %0
}

export function b $even(w %n) {
@7:
  %n.1 = call.w $abs(%n) ; @10
  %11 = eq.w %n.1, 0x0_w ; @32
  br %11, @22(0x1_b), @9
@9:
  %12 = sub.w %n.1, 0x1_w ; @33
  %b.1 = call.b $odd(%12) ; @13
  jmp @22(%b.1)
@22(%1):
  ret %1
}

export function b $odd(w %n) {
@14:
  %n.1 = call.w $abs(%n) ; @17
  %13 = eq.w %n.1, 0x0_w ; @34
  br %13, @23(0x0_b), @16
@16:
  %14 = sub.w %n.1, 0x1_w ; @35
  %b.1 = call.b $even(%14) ; @20
  jmp @23(%b.1)
@23(%2):
  ret %2
}
