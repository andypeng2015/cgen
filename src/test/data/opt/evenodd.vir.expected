module evenodd

function w $abs(w %n) {
@2:
  %3 = slt.w %n, 0x0_w ; @24
  br %3, @3, @21(%n)
@3:
  %4 = neg.w %n ; @25
  jmp @21(%4)
@21(%0):
  ret %0
}

export function b $even(w %n) {
@7:
  %n.1 = call.w $abs(%n) ; @10
  %5 = eq.w %n.1, 0x0_w ; @26
  br %5, @22(0x1_b), @9
@9:
  %6 = sub.w %n.1, 0x1_w ; @27
  %b.1 = call.b $odd(%6) ; @13
  jmp @22(%b.1)
@22(%1):
  ret %1
}

export function b $odd(w %n) {
@14:
  %n.1 = call.w $abs(%n) ; @17
  %7 = eq.w %n.1, 0x0_w ; @28
  br %7, @23(0x0_b), @16
@16:
  %8 = sub.w %n.1, 0x1_w ; @29
  %b.1 = call.b $even(%8) ; @20
  jmp @23(%b.1)
@23(%2):
  ret %2
}
